iceoryx_examples/automotive_soa/src/runtime.cpp

```diff
@@ -0,0 +1,176 @@
+{
+    iox::runtime::PoshRuntime::initRuntime(iox::RuntimeName_t(iox::cxx::TruncateToCapacity, name));
+    return GetInstance();
+}
+
+Runtime& Runtime::GetInstance() noexcept
+{
+    static Runtime runtime;
+    return runtime;
+}
+
+kom::ServiceHandleContainer<kom::ProxyHandleType>
+Runtime::FindService(kom::ServiceIdentifier& serviceIdentifier, kom::InstanceIdentifier& instanceIdentifier) noexcept
+{
+    //! [searching for iceoryx services]
+    kom::ServiceHandleContainer<kom::ProxyHandleType> iceoryxServiceContainer;
+
+    m_discovery.findService(
+        serviceIdentifier,
+        instanceIdentifier,
+        iox::cxx::nullopt,
+        [&](auto& service) {
+            iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+        },
+        iox::popo::MessagingPattern::PUB_SUB);
+
+    m_discovery.findService(
+        serviceIdentifier,
+        instanceIdentifier,
+        iox::cxx::nullopt,
+        [&](auto& service) {
+            iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+        },
+        iox::popo::MessagingPattern::REQ_RES);
+    //! [searching for iceoryx services]
+
+    // We need to make sure that all four internal services representing 'MinimalSkeleton' are available
+    //! [verify iceoryx mapping]
+    kom::ServiceHandleContainer<kom::ProxyHandleType> autosarServiceContainer;
+    if (verifyThatServiceIsComplete(iceoryxServiceContainer))
+    {
+        autosarServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+    }
+
+    return autosarServiceContainer;
+    //! [verify iceoryx mapping]
+}
+
+kom::FindServiceCallbackHandle
+Runtime::EnableFindServiceCallback(kom::FindServiceCallback<kom::ProxyHandleType> handler,
+                                   kom::ServiceIdentifier& serviceIdentifier,
+                                   kom::InstanceIdentifier& instanceIdentifier) noexcept
+{
+    // Duplicate entries for the same service are allowed
+    if (!m_callbacks.push_back(CallbackEntryType(
+            handler, {serviceIdentifier, instanceIdentifier}, NumberOfAvailableServicesOnLastSearch())))
+    {
+        std::cerr << "Callback vector capacity exceeded, did not EnableFindServiceCallback!" << std::endl;
+        std::exit(EXIT_FAILURE);
+    }
+
+    //! [attach discovery to listener]
+    if (m_callbacks.size() == 1)
+    {
+        auto invoker = iox::popo::createNotificationCallback(invokeCallback, *this);
+        m_listener.attachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, invoker)
+            .expect("Unable to attach discovery!");
+    }
+    //! [attach discovery to listener]
+
+    return kom::FindServiceCallbackHandle({serviceIdentifier, instanceIdentifier});
+}
+
+void Runtime::DisableFindServiceCallback(kom::FindServiceCallbackHandle handle) noexcept
+{
+    auto iter = m_callbacks.begin();
+    for (; iter != m_callbacks.end(); iter++)
```

Comment:
You do not use `iter` outside of this loop so and this is not the idiomatic c++ way.
You can write instead:
```cxx
auto iter = std::find_it(m_callbacks.begin(), m_callbacks.end(), 
  [&](auto &e){ return std::get<1>(e).GetServiceId() == handle.GetServiceId() && std::get<1>(e).GetInstanceId() == handle.GetInstanceId(); });
if ( iter != m_callbacks.end() ) {
  m_callbacks.erase(iter);
}
```

With this it is clear that you lookup some entry and remove it. Even better would be if the handle would have implemented an equality operator then you could use `std::find(m_callbacks.begin(), m_callbacks.end(), handle);`